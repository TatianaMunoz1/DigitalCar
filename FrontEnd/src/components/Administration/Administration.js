import React, { useState } from "react";
import { useHistory } from "react-router";
import styles from "./Administration.module.css";
import SearchCityBar from "../SearchCityBar/SearchCityBar";
import CategoryPicker from "../CategoryPicker/CategoryPicker";
import Button from "../Button/Button";
import AddFeatures from "../AddFeatures/AddFeatures";
import AddImages from "../AddImages/AddImages";
import Inputs from "../Inputs/Inputs";
import Url from "../../Url"
import axios from "axios";
import Swal from "sweetalert2";

function Administration(props) {
  const history = useHistory();
  const [newCarFeatures, setNewCarFeatures] = useState([]);
  const [newCarImages, setNewCarImages] = useState([]);
  const [submit, setSubmit] = useState(null);
  const [newCar, setNewCar] = useState({
    name: "",
    description: "",
    address: "",
    category: { id: "" },
    city: { id: "" },
  });

  const expressions = {
    input: /^[a-zA-Z0-9_.+-:", ]+$/,
  };

  const handleInputChange = (e) => {
    const target = e.target;
    const value = target.value;
    const name = target.name;

    setNewCar({
      ...newCar,
      [name]: value,
    });
  };

  const handleFilterCity = (cityId, title) => {
    setNewCar({ ...newCar, city: { id: cityId } });
  };
  const handleAddNewFeatures = (newCarFeatures) => {
    setNewCarFeatures([...newCarFeatures]);
  };
  const handleAddNewImages = (newCarImages) => {
    setNewCarImages([...newCarImages]);
  };
  const handleCategoryPicker = (categoryId) => {
    setNewCar({ ...newCar, category: { id: categoryId } });
  };

  const onSubmit = (e) => {
    e.preventDefault();
    setSubmit(true);

    if (
      newCar.name !== "" &&
      newCar.description !== "" &&
      newCar.category.id !== "" &&
      newCar.address !== "" &&
      newCar.city.id !== "" &&
      newCarFeatures.length > 0 &&
      newCarImages.length >= 4
    ) {
      const data = {
        name: `${newCar.name}`,
        description: `${newCar.description}`,
        address: newCar.address,
        features: newCarFeatures,
        category: {
          id: newCar.category.id,
        },
        images: [...newCarImages],
        city: {
          id: newCar.city.id,
        },
      };
      const config = {
        headers: { Authorization: props.user.token },
      };
      axios
        .post(Url+"/cars", data, config)
        .then(function (response) {
          if (response.status === 201) {
            Swal.fire({
              icon: "success",
              title: "Nuevo vehículo creado con exito.",
              text: "¿Desea agregar más vehículos?",
              showDenyButton: true,
              confirmButtonText: "Agregar más",
              denyButtonText: `No agregar más`,
            }).then((result) => {
              if (result.isConfirmed) {
              } else if (result.isDenied) {
                history.push("/succes-add-new-product");
              }
            });
          } else {
            const Toast = Swal.mixin({
              toast: true,
              position: "top-end",
              showConfirmButton: false,
              background: "#ffc2c2",
              timer: 5000,
              timerProgressBar: true,
              didOpen: (toast) => {
                toast.addEventListener("mouseenter", Swal.stopTimer);
                toast.addEventListener("mouseleave", Swal.resumeTimer);
              },
            });
            Toast.fire({
              icon: "error",
              title:
                "<h4 style='color:#B00020'>Lamentablemente el producto no ha podido crearse. </br>Por favor intente más tarde.</h4>",
            });
          }
        })
        .catch(function (error) {
          const Toast = Swal.mixin({
            toast: true,
            position: "top-end",
            showConfirmButton: false,
            background: "#ffc2c2",
            timer: 5000,
            timerProgressBar: true,
            didOpen: (toast) => {
              toast.addEventListener("mouseenter", Swal.stopTimer);
              toast.addEventListener("mouseleave", Swal.resumeTimer);
            },
          });
          Toast.fire({
            icon: "error",
            title:
              "<h4 style='color:#B00020'>Lamentablemente el producto no ha podido crearse. </br>Por favor intente más tarde.</h4>",
          });
        });
    }
  };

  return (
    <div className={styles.container}>
      <form className={styles.form} onSubmit={onSubmit}>
        <div>
          <h2 className={styles.title}>Datos del vehículo</h2>
          {(newCar.name === "" ||
            newCar.description === "" ||
            newCar.category.id === "" ||
            newCar.address === "" ||
            newCar.city.id === "") &&
            submit && (
              <p className={styles.error_message}>
                *Todos los campos son obligatorios. Por favor complete los
                campos vacios.
              </p>
            )}
        </div>
        <div className={styles.form_columns}>
          <div className={styles.form_first_section}>
            <div>
              <Inputs
                onChange={handleInputChange}
                value={newCar.name}
                name="name"
                type="text"
                label="Modelo del vehículo"
                errorLegend="Ingrese el módelo del vehículo."
                regularExpression={expressions.input}
              />
            </div>
            <div>
              <Inputs
                onChange={handleInputChange}
                value={newCar.address}
                name="address"
                type="text"
                label="Dirección"
                errorLegend="Ingrese la dirección de entrega del vehículo."
                regularExpression={expressions.input}
              />
            </div>
          </div>
          <div className={styles.form_second_section}>
            <div>
              <label className={styles.label}>Categoría</label>
              <CategoryPicker handleCategoryPicker={handleCategoryPicker} />
            </div>
            <div className={styles.input_city}>
              <label className={styles.label}>Ciudad</label>
              <SearchCityBar handleFilterCity={handleFilterCity} />
            </div>
          </div>
        </div>
        <div>
          <label className={styles.label}>Descripción</label>
          <textarea
            className={styles.input_textarea}
            onChange={handleInputChange}
            value={newCar.description}
            id="descripcion"
            name="description"
            rows="4"
            placeholder="Escribir aquí"
          />
        </div>
        <section>
          <h2 className={styles.title}>Agregar atributos</h2>
          {newCarFeatures.length === 0 && submit && (
            <p className={styles.error_message}>
              *Agregue al menos una característica del vehículo.
            </p>
          )}
          <AddFeatures handleAddNewFeatures={handleAddNewFeatures} />
        </section>
        <section className={styles.features}>
          <h2 className={styles.title}>Cargar imágenes</h2>
          {newCarImages.length <= 4 && submit && (
            <p className={styles.error_message}>
              *Agregue al menos 5 imagenes del vehículo.
            </p>
          )}
          <AddImages handleAddNewImages={handleAddNewImages} />
        </section>
        <div className={styles.btn_container}>
          <Button
            type="submit"
            title="Crear"
            btnStyle="btn--full"
            btnSize="btn--responsive"
            onClick={onSubmit}
          />
        </div>
      </form>
    </div>
  );
}

export default Administration;
