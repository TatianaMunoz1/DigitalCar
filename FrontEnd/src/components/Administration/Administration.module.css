.container {
  height: 100%;
}
.title {
  color: #000000;
  font-size: 20px;
  line-height: 30px;
}
.form {
  background: #f0f0f0;
  box-sizing: border-box;
  padding: 30px 15px;
}
.form,
.form_columns,
.form_first_section,
.form_second_section {
  display: flex;
  flex-direction: column;
  gap: 20px;
}
.form_first_section,
.form_second_section {
  gap: 20px;
}
.container_header {
  background: #000;
  position: absolute;
  top: 0px;
  left: 0px;
  height: 90px;
  width: 100%;
}
.error_message {
  color: rgb(255, 0, 0);
  font-size: 12px;
  font-weight: 500;
}
.label {
  display: block;
  width: 100%;
  font-weight: 500;
  font-size: 12px;
  line-height: 20px;
}
.required {
  font-size: 10.5px;
  color: #929292;
}
.input_name,
.input_direction,
.input_latitud,
.input_longitud,
.input_feature_name,
.input_feature_icon,
.input_image_title,
.input_image_url,
.input {
  height: 30px;
  width: 100%;
  padding: 6px;
  font-size: 14px;
  background: #ffffff;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
  border: none;
  border-radius: 5px;
}
.add_features_section,
.add_images_section,
.new_image_box,
.new_feature_box {
  background: #f1f1f1;
  display: flex;
  width: 100%;
  gap: 20px;
  padding: 20px;
}
.add_features,
.add_images,
.new_image,
.new_feature {
  display: grid;
  grid-template-columns: 8fr 2fr;
  gap: 20px;
  width: 100%;
}
.btn_add_feature,
.btn_add_image,
.btn_delete_image,
.btn_delete_feature {
  height: 40px;
  width: 40px;
  border-radius: 5px;
  font-size: 30px;
  line-height: 40px;
  padding: 0px;
  background: #ff7e19;
  color: #fff;
  border: none;
  align-self: flex-end;
}

.add_modify_product_policy {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  gap: 30px;
  padding: 30px;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
  border-radius: 5px;
}
.input_textarea_product_policy,
.input_textarea {
  width: 100%;
  height: 120px;
  border-radius: 8px;
  font-size: 14px;
  padding: 5px;
  border: none;
  resize: none;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
}
.input_textarea_product_policy {
  height: 400px;
}

.text_verification {
  font-size: 14px;
  color: #ff0000;
  padding-top: 5px;
}

.input_invalid {
  height: 30px;
  width: 100%;
  padding: 6px;
  font-size: 14px;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
  border: none;
  border-radius: 5px;
  border: 2px solid #ff0000;
  background: #ffe1e1;
}

.form_third_section {
  display: flex;
  flex-direction: column;
  width: 100%;
  gap: 20px;
}
.input_textarea {
  border-radius: 8px;
  font-size: 14px;
  padding: 5px;
  border: none;
  resize: none;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
}
.btn_container {
  max-width: 350px;
  width: 100%;
  align-self: center;
  padding-top: 50px;
}

/*-----TABLET-----*/
@media (min-width: 760px) and (max-width: 1280px) {
  .label {
    font-size: 14px;
  }
  .title {
    color: #000000;
    font-size: 22px;
    line-height: 30px;
  }
  .form {
    padding: 30px;
    border-radius: 0px 8px 8px 8px;
  }
  .form_first_section,
  .form_second_section {
    justify-content: space-between;
    height: 145px;
  }
  .error_message {
    font-size: 14px;
  }
  .input_name,
  .input_lastname,
  .input_mail,
  .input_city {
    padding: 0px;
    font-size: 16px;
  }
  .icon_verification {
    bottom: 12px;
    right: 10px;
  }
  .input_city_invalid {
    height: 40px;
    padding: 6px;
    border: 2px solid #ff0000;
    background: #ffe1e1;
  }
  .input_textarea,
  .container_checkbox {
    font-size: 16px;
  }
  .btn_container {
    width: 350px;
    align-self: center;
  }
}

/*-----DESKTOP-----*/
@media (min-width: 1281px) {
  .label {
    font-size: 14px;
  }
  .title {
    color: #000000;
    font-size: 24px;
    line-height: 30px;
  }
  .form {
    padding: 30px;
    border-radius: 0px 8px 8px 8px;
  }
  .form_first_section,
  .form_second_section {
    justify-content: space-between;
    height: 145px;
  }
  .form_columns {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 20px;
  }
  .form,
  .form_first_section,
  .form_second_section {
    display: flex;
    flex-direction: column;
    gap: 25px;
  }
  .input_name,
  .input_lastname,
  .input_mail,
  .input_city {
    padding: 0px;
    font-size: 16px;
  }
  .input_textarea,
  .container_checkbox {
    font-size: 16px;
  }
  .icon_verification {
    bottom: 12px;
    right: 10px;
  }
  .input_city_invalid {
    height: 40px;
    padding: 6px;
    border: 2px solid #ff0000;
    background: #ffe1e1;
  }
  .form_third_section {
    width: 100%;
    gap: 20px;
  }
  .error_message {
    font-size: 18px;
  }
  .btn_container {
    width: 350px;
    align-self: center;
  }
}
