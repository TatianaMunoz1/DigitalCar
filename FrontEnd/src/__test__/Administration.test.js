import React from "react";
import { mount, shallow } from "enzyme";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import Administration from "../components/Administration/Administration.js";

describe("Administration", () => {
  const history = createMemoryHistory();
  const wrapper = mount(
    <Router history={history}>
      <Administration />
    </Router>
  );
  test("render administration ", () => {
    const wrapper = shallow(<Administration />);
    expect(wrapper).toMatchSnapshot();
  });

  test("must be exist", () => {
    expect(wrapper.text().includes("Agregar atributos")).toBeTruthy();
    expect(wrapper.text().includes("Cargar imágenes")).toBeTruthy();
    expect(wrapper.text().includes("Categoría")).toBeTruthy();
  });
});
