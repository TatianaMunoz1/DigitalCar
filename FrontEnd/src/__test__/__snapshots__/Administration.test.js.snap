// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Administration render administration  1`] = `ShallowWrapper {}`;
